src/lib/reasoners/adt_rel.ml

```diff
@@ -218,334 +327,201 @@ let assume_th_elt t th_elt _ =
-let add_eq uf hss sm1 sm2 dep env eqs =
-  match sm1, sm2 with
-  | Adt.Alien r, Adt.Constr { c_name = h; _ }
-  | Adt.Constr { c_name = h; _ }, Adt.Alien r  ->
-    let enum, ex =
-      try MX.find r env.domains with Not_found -> hss, Ex.empty
-    in
-    let ex = Ex.union ex dep in
-    if not (HSS.mem h enum) then raise (Ex.Inconsistent (ex, env.classes));
-    let env, eqs = deduce_is_constr uf r h eqs env ex in
-    {env with domains = MX.add r (HSS.singleton h, ex) env.domains} , eqs
-
-  | Adt.Alien r1, Adt.Alien r2   ->
-    let enum1,ex1 =
-      try MX.find r1 env.domains with Not_found -> hss, Ex.empty in
-    let enum2,ex2 =
-      try MX.find r2 env.domains with Not_found -> hss, Ex.empty in
-    let ex = Ex.union dep (Ex.union ex1 ex2) in
-    let diff = HSS.inter enum1 enum2 in
-    if HSS.is_empty diff then raise (Ex.Inconsistent (ex, env.classes));
-    let domains = MX.add r1 (diff, ex) env.domains in
-    let env = {env with domains = MX.add r2 (diff, ex) domains } in
-    if HSS.cardinal diff = 1 then begin
-      let h' = HSS.choose diff in
-      let env, eqs = deduce_is_constr uf r1 h' eqs env ex in
-      let env, eqs = deduce_is_constr uf r2 h' eqs env ex in
-      env, eqs
-    end
-    else env, eqs
+  | Alien r ->
+    begin match X.type_info r with
+      | Ty.Tadt (name, params) as ty ->
+        let Ty.Adt body = Ty.type_body name params in
+        let ds =
+          try Ty.assoc_destrs c body with Not_found -> assert false
+        in
+        let xs = List.map (fun (_, ty) -> E.fresh_name ty) ds in
+        Some (r, E.mk_constr (Hstring.view c) xs ty)
 
-  |  _ -> env, eqs
+      | _ -> assert false
+    end
 
+  | Constr _ ->
+    (* The semantic value [r] is already a constructor application, so
+       we don't need to produce a new equation in this case. *)
+    None
 
-let add_aux env r =
-  Debug.add r;
-  match Th.embed r with
-  | Adt.Alien r when not (MX.mem r env.domains) ->
-    begin match values_of (X.type_info r) with
-      | Some s ->
-        (* All the constructors are possible. *)
-        { env with domains = MX.add r (s, Ex.empty) env.domains }
-      | None ->
-        env
-    end
-  | _ -> env
+  | Select _ ->
+    assert false
 
-(* Needed for models generation because fresh terms are not
-   added with function add. *)
-let add_rec env r = List.fold_left add_aux env (X.leaves r)
+let propagate_domains env =
+  Domains.propagate
+    (fun (eqs, new_terms) rr d ->
+       match Domain.as_singleton d with
+       | Some (c, ex) ->
+         begin match build_constr_eq rr c with
+           | Some (r, cons) ->
+             let r', ctx = X.make cons in
+             (* In the current implementation of `X.make`, we produce
+                a nonempty context only for interpreted semantic values
+                of the `Arith` and `Records` theories. The semantic
+                values `cons` never involves such values. *)
+             assert (Lists.is_empty ctx);
```

Comment:
Ideally the call to `X.make` should be called at the same point that the term is created (otherwise we might change the code in `build_constr_eq` and break this property accidentally). Could `build_constr_eq` call `X.make` and return `Some (r, cons, r')`?